doc/SAI-Proposal-SAI-PTF.md

```diff
@@ -0,0 +1,132 @@
+
+
+All of this can be generated with gensairpc.pl script.
+
+Autogeneration:
+
+```
+cd <root>/sai/
+./gensairpc.pl --clean
+```
+
+When upgrading SAI:
+
+```
+cd SAI/meta
+make
+cp saimetadata.c <root>/sai
+cp saimetadata.h <root>/sai
+cd <root>/sai
+./gensairpc.pl --clean --no-meta-build
+make clean - C SAI/meta
+```
+
+Use --no-meta-build because SAI meta are already built with make.
+
+## Test execution
+
+Test files written in Python call functions from sai_adapter.py. When running such test auto-generated Python functions are talking with C-based RPC server to run C functions. These functions are used to create/remove SAI object and get/set their attributes. Then it is running internal SAI implementation adopting common SAI interface to test the implementation.
+
+Tests use SAI to configure underlay switch. Using PTF we send and verify packets that are testing each switch configuration.
+
+## Tests structure
+
+The tests themselves are based on python unittest framework (https://docs.python.org/3/library/unittest.html). They consist of three main functions:
+
+- setUp() - creating necessary     objects
+- runTest() - running proper test
+- tearDown() - removing all     created objects
+
+Tests have to inherit from one of two base classes - SaiHelper or SaiHelperBase. SaiHelperBase is the base class that runs initial configuration for all tests. SaiHelper inherits from SaiHelperBase but provides additional configuration.
+
+The based configuration created by SaiHelper looks like this:
+
+| Name       | Vlan ID | Ports | Tagging |
+| ----------- | ------------ | ------------ | ------------ |
+| vlan10    | 10  | port0 <br />port1 <br />lag1| untagged <br />tagged <br />untagged |
+| vlan20    | 20  | port2 <br />port3 <br />lag2| untagged <br />tagged <br />tagged |
+| lag1   | --  | port4 <br />port5 <br />port6| -- |
+| lag2   | --  | port7 <br />port8 <br />port9| -- |
+| port10_rif   | --  | port10 | -- |
+| port11_rif   | --  | port11 | -- |
+| port12_rif   | --  | port12 | -- |
+| port13_rif   | --  | port13 | -- |
+| lag3 <br />lag3_rif   | --  | port14 <br />port15 <br />port16| -- |
+| lag4 <br />lag4_rif   | --  | port17 <br />port18 <br />port19| -- |
+| vlan30 <br />vlan30_rif    | 30  | port20 <br />port21 <br />lag5| untagged <br />tagged <br />tagged |
+| lag5   | --  | port22 <br />port23 | -- |
+| Additional ports   | --  | port24 <br />port25 <br />port26<br />port27<br />port28<br />port29<br />port30<br />port31| -- |
+
+Ports 24-31 can be used by every test to create additional objects.
+
+Both SaiHelper and SaiHelperBase classes are created in sai_base_test.py file.
+
+If tests intend to have more cmplicated setUp then the one that is created in the base class they need to imlpement setUp method but remember to call the setUp from the base class first.
+
+## Implemented tests
+
+Tests are divided into per-feature files such as saivlan.py, saifdb.py etc. Framework contains tests for following features:
```

Comment:
In overview section we have "Apart from the framework there's a set of tests". Could we use another words instead "Framework contains" here?